File: src/test/java/bankaccount/AccountTest.java
package bankaccount;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

import java.time.LocalDate;
import java.time.Month;

import static org.assertj.core.api.Assertions.assertThat;

public class AccountTest {
    private Account account;
    private Printer printer;
    private DateService dateService;

    @Before
    public void setUp() {
        dateService = Mockito.mock(DateService.class);
        account = new Account(dateService);
        printer = Mockito.mock(Printer.class);
    }

    @Test
    public void balance_should_be_zero_by_default() {
        assertThat(account.balance()).isEqualTo(Amount.ZERO);
    }

    @Test(expected = IllegalArgumentException.class)
    public void deposit_should_not_allow_negative_amounts() {
        account.deposit(Amount.of(-10));
    }

    @Test
    public void deposit_should_update_account_balance() {
        account.deposit(Amount.of(10));

        assertThat(account.balance()).isEqualTo(Amount.of(10));
    }

    @Test
    public void multiple_deposit_should_update_account_balance() {
        account.deposit(Amount.of(10));
        account.deposit(Amount.of(10));

        assertThat(account.balance()).isEqualTo(Amount.of(20));
    }

    @Test
    public void multiple_deposit_of_different_amount_should_update_balance() {
        account.deposit(Amount.of(10));
        account.deposit(Amount.of(15));

        assertThat(account.balance()).isEqualTo(Amount.of(25));
    }

    @Test
    public void withdraw_nothing_should_not_update_balance() {
        account.withdraw(Amount.ZERO);

        assertThat(account.balance()).isEqualTo(Amount.ZERO);
    }

    @Test(expected = IllegalArgumentException.class)
    public void withdraw_should_not_allow_negative_amounts() {
        account.withdraw(Amount.of(-10));
    }

    @Test
    public void withdraw_should_update_account_balance() {
        account.withdraw(Amount.of(50));

        assertThat(account.balance()).isEqualTo(Amount.of(-50));
    }

    @Test
    public void multiple_withdraw_should_update_account_balance() {
        account.withdraw(Amount.of(50));
        account.withdraw(Amount.of(50));

        assertThat(account.balance()).isEqualTo(Amount.of(-100));
    }

    @Test
    public void history_should_print_a_description_line() {
        account.history(printer);

        Mockito.verify(printer).print("OPERATION | DATE | AMOUNT | BALANCE");
    }

    @Test
    public void history_should_keep_track_of_deposits() {
        Mockito.when(dateService.now())
                .thenReturn(LocalDate.of(2017, Month.AUGUST, 27));
        account.deposit(Amount.of(1000));
        account.history(printer);

        Mockito.verify(printer).print("DEPOSIT | 27/08/2017 | +10€ | +10€");
    }

    @Test
    public void history_should_keep_track_of_withdrawals() {
        Mockito.when(dateService.now())
                .thenReturn(LocalDate.of(2017, Month.AUGUST, 27));
        account.withdraw(Amount.of(1000));
        account.history(printer);

        Mockito.verify(printer).print("WITHDRAW | 27/08/2017 | +10€ | -10€");
    }
}

File: src/test/kotlin/bankaccount/bankaccount/AccountTest.kt
package bankaccount.bankaccount

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import java.time.LocalDate
import java.time.Month
import java.time.format.DateTimeFormatter

class AccountTest {
    private lateinit var account: Account
    private lateinit var dateService: DateService
    @BeforeEach
    fun setUp() {
        dateService = Mockito.mock(DateService::class.java)
        account = Account(dateService)

        val formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy")
        val date = LocalDate.parse("29-06-2023", formatter)

        Mockito.`when`(dateService.now()).thenReturn(date)
    }

    @Test
    fun balance_should_be_zero_by_default() {
        Assertions.assertThat(account.balance()).isEqualTo(Amount.ZERO)
    }

    @Test
    fun deposit_should_not_allow_negative_amounts() {
        Assertions.assertThatThrownBy { account.deposit(Amount.of(-10)) }.isInstanceOf(
            IllegalArgumentException::class.java
        )
    }

    @Test
    fun deposit_should_update_account_balance() {
        account.deposit(Amount.of(10))
        Assertions.assertThat(account.balance()).isEqualTo(Amount.of(10))
    }

    @Test
    fun multiple_deposit_should_update_account_balance() {
        account.deposit(Amount.of(10))
        account.deposit(Amount.of(10))
        Assertions.assertThat(account.balance()).isEqualTo(Amount.of(20))
    }

    @Test
    fun multiple_deposit_of_different_amount_should_update_balance() {
        account.deposit(Amount.of(10))
        account.deposit(Amount.of(15))
        Assertions.assertThat(account.balance()).isEqualTo(Amount.of(25))
    }

    @Test
    fun withdraw_nothing_should_not_update_balance() {
        account.withdraw(Amount.ZERO)
        Assertions.assertThat(account.balance()).isEqualTo(Amount.ZERO)
    }

    @Test
    fun withdraw_should_not_allow_negative_amounts() {
        Assertions.assertThatThrownBy { account.withdraw(Amount.of(-10)) }.isInstanceOf(
            IllegalArgumentException::class.java
        )
    }

    @Test
    fun withdraw_should_update_account_balance() {
        account.withdraw(Amount.of(50))
        Assertions.assertThat(account.balance()).isEqualTo(Amount.of(-50))
    }

    @Test
    fun multiple_withdraw_should_update_account_balance() {
        account.withdraw(Amount.of(50))
        account.withdraw(Amount.of(50))
        Assertions.assertThat(account.balance()).isEqualTo(Amount.of(-100))
    }

    @Test
    fun history_should_return_an_empty_list_when_empty() {
        val history = account.history()
        Assertions.assertThat(history).isEmpty()
    }

    @Test
    fun history_should_keep_track_of_deposits() {
        val today = LocalDate.of(2017, Month.AUGUST, 27)
        Mockito.`when`(dateService.now())
            .thenReturn(today)
        account.deposit(Amount.of(1000))

        val history = account.history()

        val expectedLine = DepositLine.of(
            Amount.of(1000), today, Amount.of(1000)
        )
        Assertions.assertThat(history).containsExactly(expectedLine)
    }

    @Test
    fun history_should_keep_track_of_withdrawals() {
        Mockito.`when`(dateService.now())
            .thenReturn(LocalDate.of(2017, Month.AUGUST, 27))
        account.withdraw(Amount.of(1000))

        val history = account.history()

        val expectedLine = WithdrawLine.of(Amount.of(1000), LocalDate.of(2017, Month.AUGUST, 27), Amount.of(1000))
        Assertions.assertThat(history).containsExactly(expectedLine)
    }
}
